experimental/oak_async/tests/tests.rs

```diff
@@ -0,0 +1,287 @@
+    });
+
+    let result = block_on(async {
+        let a: DummyData = do_channel_read(HANDLE_0)
+            .await
+            .expect("Failed to read channel HANDLE_0");
+        let b: DummyData = do_channel_read(HANDLE_1)
+            .await
+            .expect("Failed to read channel HANDLE_1");
+
+        [a.0, b.0].join(", ")
+    })
+    .expect("block_on failed");
+
+    assert_eq!(result, "Hello, world!");
+}
+
+#[test]
+fn block_on_multiple_readers_parallel() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        for handle in handles.iter_mut() {
+            let data = match handle.handle() {
+                HANDLE_0 => "Hello",
+                HANDLE_1 => "world!",
+                c => panic!("Unexpected waiting channel {}", c),
+            };
+            add_ready_data(handle.handle(), &DummyData::new(data));
+            handle.set_status(ChannelReadStatus::ReadReady);
+        }
+        OakStatus::Ok
+    });
+
+    let (a, b): (Result<DummyData, OakError>, Result<DummyData, OakError>) = block_on(
+        futures::future::join(do_channel_read(HANDLE_0), do_channel_read(HANDLE_1)),
+    )
+    .expect("block_on failed");
+
+    assert_eq!(
+        a.expect("Read from channel HANDLE_0 failed"),
+        DummyData::new("Hello")
+    );
+
+    assert_eq!(
+        b.expect("Read from channel HANDLE_1 failed"),
+        DummyData::new("world!")
+    );
+}
+
+#[test]
+fn block_on_drop_channel_read_immediately() {
+    init();
+    block_on(async {
+        // The future is immediately dropped, so never polled
+        let _ = do_channel_read::<DummyData>(1);
+    })
+    .expect("block_on failed");
+}
+
+#[test]
+fn block_on_drop_channel_read_after_wait() {
+    init();
+    set_wait_on_channels_handler(|handles| {
+        std::thread::sleep(std::time::Duration::from_secs(1));
+        for handle in handles.iter_mut() {
+            match handle.handle() {
+                HANDLE_0 => {
+                    handle.set_status(ChannelReadStatus::NotReady);
+                }
+                HANDLE_1 => {
+                    add_ready_data(HANDLE_1, &DummyData::new("hello"));
+                    handle.set_status(ChannelReadStatus::ReadReady);
+                }
+                c => panic!("Unexpected waiting channel {}", c),
+            };
+        }
+        OakStatus::Ok
+    });
+
+    let result: Result<DummyData, OakError> = block_on(async {
+        futures::future::select(do_channel_read(HANDLE_0), do_channel_read(HANDLE_1))
+            .await
+            .factor_first()
+            // Note: It is important that we extract the result inside the async block. Doing this
+            // outside the async scope will block indefinitely, because it will try and move the
+            // unresolved ChannelRead on HANDLE_0 outside of the async block, but for block_on to
+            // terminate that ChannelRead must be `Drop`ed first (to remove it from the waiting
+            // set).
```

Comment:
Uhmm this is not ideal, is there anything that can be done to catch it in a more programmatic way?

Reply:
It would be nice if there was a way to express a concept like 'this function can return any type except for X', but we don't have that at the moment. I don't see a great way of enforcing this programmatically, but on the flip side we won't be exposing `block_on` to user code in the final version, instead we'll use it to implement the async event loop, which returns nothing and thus doesn't suffer from the same issue.